cpp/src/arrow/chunked_array.h

```diff
@@ -182,6 +183,7 @@ class ARROW_EXPORT ChunkedArray {
   std::shared_ptr<DataType> type_;
 
  private:
+  std::unique_ptr<internal::ChunkResolver> chunk_resolver_;
```

Comment:
Is this useful for this to be dynamically allocated, since the constructor always initializes it? If this is so as to defer initialization, this could use `util::optional` instead (but `ChunkResolver` could simply have a default constructor).

Reply:
Again, you can easily define a move constructor for `ChunkResolver` just as you did for the copy constructor, so I'm not sure what the actual difficulty is.
Something like:
```c++
  ChunkResolver(ChunkResolver&& other)
    : offsets_(std::move(other.offsets_)),
      cached_chunk_(other.cached_chunk_.load()) {}
```

Reply:
Added move constructor and move assignment operator, so now we can instantiate a non-dynamic member variable.